pkg/gateway/multipart/iterator.go

```diff
@@ -0,0 +1,110 @@
+package multipart
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/treeverse/lakefs/pkg/kv"
+)
+
+// UploadIterator is an iterator over multipart uploads sorted by Path, then UploadID
+type UploadIterator interface {
+	// Next advances the iterator to the next upload
+	// Returns true if there is a next upload, false otherwise
+	Next() bool
+	// Value returns the current upload
+	// Should only be called after Next returns true
+	Value() *Upload
+	// Err returns any error encountered during iteration
+	Err() error
+	// Close releases resources associated with the iterator
+	Close()
+	// SeekGE seeks to the first upload with Path > key, or if Path == key, UploadID > uploadID
+	// After calling SeekGE, Next() must be called to access the first element at or after the seek position
+	SeekGE(key string, uploadID string)
+}
+
+// sortedUploadIterator collects all uploads, sorts them, and provides efficient seeking
+type sortedUploadIterator struct {
+	uploads []*Upload
+	current int
+	err     error
+}
+
+// newSortedUploadIterator creates a sorted iterator from the kv store
+// It loads all uploads into memory and sorts them by Path, then UploadID
+func newSortedUploadIterator(ctx context.Context, store kv.Store, partitionKey string) (*sortedUploadIterator, error) {
+	kvIter := kv.NewPartitionIterator(ctx, store, (&UploadData{}).ProtoReflect().Type(), partitionKey, 0)
+	defer kvIter.Close()
+
+	var uploads []*Upload
+	for kvIter.Next() {
+		entry := kvIter.Entry()
+		if entry == nil {
+			continue
+		}
+		data, ok := entry.Value.(*UploadData)
+		if !ok {
+			return nil, fmt.Errorf("unexpected value type: %T", entry.Value)
+		}
+		uploads = append(uploads, multipartFromProto(data))
+	}
+	if err := kvIter.Err(); err != nil {
+		return nil, err
+	}
```

Comment:
If I'm reading this correctly - you're fetching all upload IDs from the kv store to return a single page. I understand how it solves the problem but I'm afraid this doesn't scale. What if I have 10M UploadIDs?

Reply:
Can we saved them in the kv store by their path and then upload id? It will give us what we need and also scales